src/PostgREST/Error.hs

```diff
@@ -143,166 +148,159 @@ instance PgrstError ApiRequestError where
-  , "hint"     .= hint
-  ]
-
-instance JSON.ToJSON ApiRequestError where
-  toJSON (QueryParamError (QPError message details)) = toJsonPgrstError
-    ApiRequestErrorCode00 message (Just (JSON.String details)) Nothing
-
-  toJSON (InvalidRpcMethod method) = toJsonPgrstError
-    ApiRequestErrorCode01 ("Cannot use the " <> T.decodeUtf8 method <> " method on RPC") Nothing Nothing
-
-  toJSON (InvalidBody errorMessage) = toJsonPgrstError
-    ApiRequestErrorCode02 (T.decodeUtf8 errorMessage) Nothing Nothing
-
-  toJSON (InvalidRange rangeError) = toJsonPgrstError
-    ApiRequestErrorCode03
-    "Requested range not satisfiable"
-    (Just $ case rangeError of
+-- Error codes:
+--
+-- Error codes are grouped by common modules or characteristics
+-- New group of errors will be added at the end of all the groups and will have the next prefix in the sequence
+-- Keep the "PGRST" prefix in every code for an easier search/grep
+-- They are grouped as following:
+--
+-- PGRST0xx -> Connection Error
+-- PGRST1xx -> ApiRequest Error
+-- PGRST2xx -> SchemaCache Error
+-- PGRST3xx -> JWT authentication Error
+-- PGRSTXxx -> Internal Hasql Error
+
+instance ErrorBody ApiRequestError where
+  -- CODE: Text
+  code QueryParamError{}           = "PGRST100"
+  code InvalidRpcMethod{}          = "PGRST101"
+  code InvalidBody{}               = "PGRST102"
+  code InvalidRange{}              = "PGRST103"
+  -- code ParseRequestError           = "PGRST104" -- no longer used
+  code InvalidFilters              = "PGRST105"
+  code UnacceptableSchema{}        = "PGRST106"
+  code MediaTypeError{}            = "PGRST107"
+  code NotEmbedded{}               = "PGRST108"
+  -- code LimitNoOrderError           = "PGRST109" -- no longer used
+  -- code OffLimitsChangesError       = "PGRST110" -- no longer used
+  code GucHeadersError             = "PGRST111"
+  code GucStatusError              = "PGRST112"
+  -- code BinaryFieldError            = "PGRST113" -- no longer used
+  code PutLimitNotAllowedError     = "PGRST114"
+  code PutMatchingPkError          = "PGRST115"
+  code SingularityError{}          = "PGRST116"
+  code UnsupportedMethod{}         = "PGRST117"
+  code RelatedOrderNotToOne{}      = "PGRST118"
+  -- code SpreadNotToOne              = "PGRST109" -- no longer used
+  code UnacceptableFilter{}        = "PGRST120"
+  code PGRSTParseError{}           = "PGRST121"
+  code InvalidPreferences{}        = "PGRST122"
+  code AggregatesNotAllowed        = "PGRST123"
+  code MaxAffectedViolationError{} = "PGRST124"
+  code InvalidResourcePath         = "PGRST125"
+  code OpenAPIDisabled             = "PGRST126"
+  code NotImplemented{}            = "PGRST127"
+
+  code NoRelBetween{}              = "PGRST200"
+  code AmbiguousRelBetween{}       = "PGRST201"
+  code NoRpc{}                     = "PGRST202"
+  code AmbiguousRpc{}              = "PGRST203"
+  code ColumnNotFound{}            = "PGRST204"
+  code TableNotFound{}             = "PGRST205"
```

Comment:
This is great work, things are already looking much better :tada: 

Here I was wondering if we should have a type besides `ApiRequestError` for these schema cache errors? That way the group prefixes are clearly separated?

Reply:
Ok, cool. Let's leave that for later.